components/users/api/src/main/java/org/phenotips/groups/internal/UsersAndGroups.java

```diff
@@ -0,0 +1,249 @@
+
+import com.xpn.xwiki.XWiki;
+import com.xpn.xwiki.XWikiContext;
+import com.xpn.xwiki.doc.XWikiDocument;
+
+/**
+ * @version $Id$
+ */
+@Component(roles = { UsersAndGroups.class })
+@Singleton
+public class UsersAndGroups
+{
+    /** Result returned in case getType() result is a user. */
+    public static final String USER = "user";
+
+    /** Result returned in case getType() result is a group. */
+    public static final String GROUP = "group";
+
+    /** Result returned in case getType() result is unknown. */
+    public static final String UNKNOWN = "unknown";
+
+    private static final EntityReference USER_CLASS = new EntityReference("XWikiUsers", EntityType.DOCUMENT,
+        new EntityReference(XWiki.SYSTEM_SPACE, EntityType.SPACE));
+
+    private static final EntityReference GROUP_CLASS = new EntityReference("XWikiGroups", EntityType.DOCUMENT,
+        new EntityReference(XWiki.SYSTEM_SPACE, EntityType.SPACE));
+
+    private static final String INPUT_PARAMETER = "input";
+
+    private static final String INPUT_FORMAT = "%%%s%%";
+
+    private static final String ID_KEY = "id";
+
+    private static final String VALUE_KEY = "value";
+
+    private static final String ICON_KEY = "icon";
+
+    private static String usersQueryString;
+
+    private static String groupsQueryString;
+
+    @Inject
+    private QueryManager qm;
+
+    @Inject
+    private Logger logger;
+
+    @Inject
+    private UserManager userManager;
+
+    @Inject
+    private GroupManager groupManager;
+
+    @Inject
+    private DocumentAccessBridge bridge;
+
+    @Inject
+    private Provider<XWikiContext> xcontextProvider;
+
+    static {
+        StringBuilder usersQuerySb = new StringBuilder();
+        usersQuerySb.append("from doc.object(XWiki.XWikiUsers) as user ");
+        usersQuerySb.append(" where lower(doc.name) like :").append(UsersAndGroups.INPUT_PARAMETER);
+        usersQuerySb.append(" or concat(concat(lower(user.first_name), ' '), lower(user.last_name)) like ");
+        usersQuerySb.append(":").append(UsersAndGroups.INPUT_PARAMETER);
+        usersQuerySb.append(" order by user.first_name, user.last_name");
+        UsersAndGroups.usersQueryString = usersQuerySb.toString();
+
+        StringBuilder groupsQuerySb = new StringBuilder();
+        groupsQuerySb.append("from doc.object(PhenoTips.PhenoTipsGroupClass) as groups ");
+        groupsQuerySb.append(" where lower(doc.name)  like :").append(UsersAndGroups.INPUT_PARAMETER);
+        groupsQuerySb.append(" and doc.fullName <> 'PhenoTips.PhenoTipsGroupTemplate' ");
+        groupsQuerySb.append(" order by doc.name");
+        UsersAndGroups.groupsQueryString = groupsQuerySb.toString();
+    }
+
+    /**
+     * Checks whether an {@link userOrGroup} is an entity that represents a user or a group.
+     *
+     * @param userOrGroup entity to check
+     * @return {@link USER}, @{link GROUP}, or {@link UNKNOWN}
+     */
+    public String getType(EntityReference userOrGroup)
```

Comment:
my reasoning was that since the return value is not directly usable (it is a new constant, so can't just use return type as a parameter to other code) the code that uses this would have to have a pattern like:

```var type = getType()
if (type == A) ... else if (type == B) .... ```
(or `if (getType() == A) ... else if (getType() == B) ...` which is less efficient)

which can be replaced by more obvious non-custom constant `if (isUser()) ... else if (isGroup()) ...`